README.md
# Sparkledriver

![Sparkledriver glitter tractor](https://raw.githubusercontent.com/jackrusher/sparkledriver/master/assets/SparkleDriver.jpg)

A mostly idiomatic wrapper
around
[jBrowserDriver](https://github.com/MachinePublishers/jBrowserDriver),
which is a pure-Java [Selenium](http://seleniumhq.org/)-compatible
interface to the [WebKit](https://webkit.org) browser library included
in recent JVM versions (>=1.8) as part
of
[JavaFX](http://docs.oracle.com/javase/8/javafx/get-started-tutorial/jfx-overview.htm#JFXST784).

This library is useful when you need a full-featured browser with
support for cookies, Javascript, and so on, with no native
dependencies. I've found it to be a nice basis for testing web
applications, scraping content from Javascript-heavy websites,
automating interactions with services, and so on.

**N.B.** While Oracle's JVM bundles JavaFX, the OpenJDK JVM does not
come with it. If you're using OpenJDK (which is the default on
_ubuntu_, for example), you must also install the OpenJFX package on
your chosen platform in order to use this library.

## Quick Start

``` clojure
[sparkledriver "0.2.4"]
```

This example fetches the Clojure homepage, extracts some elements and
retrieves the text of one:

``` clojure
(require '[sparkledriver.browser :refer [with-browser make-browser fetch!]])
(require '[sparkledriver.element :refer [find-by-xpath* text]])

(with-browser [browser (make-browser)]
  (-> (fetch! browser "http://clojure.org")
      (find-by-xpath* "//div[@class='clj-intro-message']/p")
      (nth 2)
      text))
;;=> "I hope you find Clojure's combination of facilities elegant, powerful, practical and fun to use."
```

There are many more specific examples in the test suite. In additon,
the automatically
generated [codox](https://github.com/weavejester/codox) documentation
is [here](https://jackrusher.github.io/sparkledriver/).

## A Note on Dependencies

A case has
been [reported](https://github.com/jackrusher/sparkledriver/issues/5)
where an old version of `common-codec`, as a library dependency, was
causing the tests to fail. If you experience this issue, please add
the following dependency:

```clj
[commons-codec "1.9"]
```

to your project.

## Completeness/Robustness

While there are still a few areas where the underlying APIs could be
more completely wrapped, the core functionality has been enrobed in
delicious parentheses. Some bits are better tested than others, but I
*am* currently using this library in production.

## License

Copyright © 2016-2018 Jack Rusher

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Added
### Changed

## [0.2.4] - 2018-11-22
### Changed
- Upgrade to jBrowserDriver `1.0.1`

## [0.2.3] - 2018-08-23
### Changed
- Upgrade to jBrowserDriver `1.0.0`
- Tested against clojure `1.9.0`

## [0.2.2] - 2017-11-06
### Changed
- Upgrade to jBrowserDriver `0.17.11`

## [0.2.1] - 2017-09-29
### Added
- `(make-browser :block-ads true)` now creates a browser instance that
  blocks
  ads. [Thanks](https://github.com/jackrusher/sparkledriver/pull/14),
  Andreas.
### Changed
- Documentation and tests improved.
### Fixed
- [Issue](https://github.com/jackrusher/sparkledriver/issues/17) where
  sparkledriver would hang if `close-browser!` was called while (a)
  more than one window was open, and (b) the current focus wasn't the
  first window.

## [0.2.0] - 2017-08-29
### Added
- `sparkledriver.retry` namespace with code supporting automatic
  retries
- general improvement of documentation (docstrings and emitted codox
  output)
### Changed
- Reorganized the code into multiple namespaces, by function.
- element finding functions use the `*reply-fn*` dynamic variable to
  specify a retry handler function.
- `dismiss-alert` is now `dismiss-alert!`
- `accept-alert` is now `accept-alert!`

## [0.1.10] - 2017-08-16
### Added
- `page-text` returns the textual content of the current page
- `title` returns the page's title
- `set-cookie!` set a cookie
- `delete-cookies!` delete a specific cookie
- `delete-all-cookies!` clear out the cookie jar

### Changed
- `fetch!` now calls `str` on its argument, making it compatible with URL types like `java.net.URL` and `lambdaisland.uri.URI`.

## [0.1.9] - 2017-08-09
### Added
- Functions to return an element's `outer-html`.
- Tests for `inner-html` and `outer-html`.
### Changed
- Upgrade to jBrowserDriver `0.17.9`

## [0.1.8] - 2017-06-24
### Added
- Functions to evaluate JS in the context of the current page:
  `execute-script` and `execute-script-async`.
### Changed
- New all-local test harness no longer requires Internet access.

## [0.1.7] - 2017-06-19
### Added
- functions for interacting with `alert`s (`switch-to-alert`,
  `accept-alert`, `dismiss-alert`) and multiple windows
  (`current-window`, `all-windows`, `switch-to-window`,
  `maximize-window`).

## [0.1.6] - 2017-06-08
### Added
- `:screen-size [width height]` browser builder option to change the
  window size of the virtual browser.
### Changed
- jBrowserDriver version 0.17.8
- default browser window size is now 1366 x 768

## [0.1.5] - 2017-04-20
### Added
- page-wait
- attachments-dir
- media-dir
- cache-dir

## [0.1.4] - 2017-04-19
### Fixed
- terrible error where code wouldn't compile

## [0.1.3] - 2017-04-19
### Added
- Parameterized browser options
- Browser log access functions (wire tracing, for example)

### Changed
- jBrowserDriver version 0.17.7

## [0.1.2] - 2017-04-03
### Added
- Finished the wrapper for WebElement.

### Changed
- jBrowserDriver version 0.17.6.
- Improved docstrings

## [0.1.1] - 2016-11-04
- Initial release

[Unreleased]: https://github.com/your-name/sparkledriver/compare/0.2.1...HEAD
[0.2.1]: https://github.com/your-name/sparkledriver/compare/0.2.1...0.2.1
[0.2.0]: https://github.com/your-name/sparkledriver/compare/0.1.9...0.2.0
[0.1.9]: https://github.com/your-name/sparkledriver/compare/0.1.8...0.1.9
[0.1.8]: https://github.com/your-name/sparkledriver/compare/0.1.7...0.1.8
[0.1.7]: https://github.com/your-name/sparkledriver/compare/0.1.6...0.1.7
[0.1.6]: https://github.com/your-name/sparkledriver/compare/0.1.5...0.1.6
[0.1.5]: https://github.com/your-name/sparkledriver/compare/0.1.4...0.1.5
[0.1.4]: https://github.com/your-name/sparkledriver/compare/0.1.3...0.1.4
[0.1.3]: https://github.com/your-name/sparkledriver/compare/0.1.2...0.1.3
[0.1.2]: https://github.com/your-name/sparkledriver/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/your-name/sparkledriver/compare/0.1.0...0.1.1


test/sparkledriver/core_test.clj
(ns sparkledriver.core-test
  (:require [clojure.test :refer :all]
            [sparkledriver.browser :refer :all]
            [sparkledriver.element :refer :all]
            [sparkledriver.cookies :refer :all]
            [sparkledriver.retry :refer :all]
            [org.httpkit.server :refer [run-server]]
            [compojure.core :refer :all]
            [compojure.route :as route]
            [hiccup.core :refer [html]])
  (:import java.net.URI))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; setup a local webserver and test page

(def port 9999)
(declare start-server stop-server)

(use-fixtures :once (fn [tests]
                      (start-server port)
                      (tests)
                      (stop-server)))

(defonce server-handle (atom nil))

(def test-page
  [:html
   [:head
    [:title "The Second Coming"]
    [:meta {:name "turning and turning" :content "in the widening gyre"}]
    [:meta {:name "the falcon" :content "cannot hear the falconer"}]
    [:meta {:name "things fall apart" :content "the centre cannot hold"}]]
   [:body
    [:div
     [:h1#heroic "Sparkledriver, driven"]
     [:div.emptiness]
     [:p [:span {:data-direction "slouching towards bethlehem"}
          "And what rough beast, its hour come round at last"]]
     [:div.emptiness]
     [:p.vast-image "Spiritus Mundi"]
     [:div.emptiness]]
    [:div
     [:p#loosed-upon-the-world.some-revelation {:style "color:red;"}
      [:span.mere-anarchy "The blood-dimmed tide is loosed, and everywhere"]
      [:span.mere-anarchy "The ceremony of innocence is drowned"]]]
    [:div
     [:img#test-ad {:src "data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="}]]]])

(defroutes taxi-server
  (GET "/" [] (html test-page))
  (route/not-found "<h1>Page not found</h1>"))

(defn stop-server []
  (when @server-handle (@server-handle)))

(defn start-server [port]
  (swap! server-handle
         (fn [_]
           (stop-server) ;; kill the old server first, if needed
           (run-server taxi-server {:port port}))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; tests against the local test page

(deftest basic-tests
  (with-browser [browser (fetch! (make-browser) (str "http://0.0.0.0:" port))]
    (testing "loads the page"
      (is (= 200 (status-code browser))))

    (testing "tag selectors"
      (is (= (-> (find-by-tag browser "h1") text)
             "Sparkledriver, driven"))
      (is (= (-> (find-by-tag browser "span") (attr "data-direction"))
             "slouching towards bethlehem"))
      (is (= (sort (mapv #(attr % "name") (find-by-tag* browser "meta")))
             '("the falcon" "things fall apart" "turning and turning"))))
    
    (testing "class selectors"
      (is (= (text (find-by-class browser "vast-image"))
             "Spiritus Mundi"))
      (is (= (distinct (map tag (find-by-class* browser "emptiness")))
             '("div"))))
    
    (testing "css selectors"
      (is (= (mapv text (find-by-css* browser "span.mere-anarchy"))
             ["The blood-dimmed tide is loosed, and everywhere"
              "The ceremony of innocence is drowned"])))

    (testing "id selectors"
      (is (= (text (find-by-id browser "heroic"))
             "Sparkledriver, driven"))
      (is (= (css-value (find-by-id browser "loosed-upon-the-world") "color")
             "rgba(255, 0, 0, 1)")))

    (testing "xpath selectors"
      (is (= (-> (find-by-xpath browser "//h1") text)
             "Sparkledriver, driven"))
      (is (= (-> (find-by-xpath* browser "//p[@class='some-revelation']/span")
                 (second)
                 text)
             "The ceremony of innocence is drowned")))

    (testing "inner/outer html"
      (is (= (-> (find-by-xpath browser "//h1") outer-html)
             "<h1 id=\"heroic\">Sparkledriver, driven</h1>"))
      (is (= (-> (find-by-xpath browser "//h1") inner-html)
             "Sparkledriver, driven")))
    
    (testing "javascript execution"
      ;; use a script to set an attribute
      (execute-script browser "document.getElementById(\"heroic\")[\"data-villainous\"] = \"frenemy\";")
      (is (= (attr (find-by-id browser "heroic") "data-villainous")
             "frenemy"))
      ;; pass an element as an argument to javascript, receive it in return, get an attr from it
      (is (= (attr (execute-script browser "return arguments[0];", (find-by-id browser "heroic")) "data-villainous")
             "frenemy")))

    (testing "page helpers"
      (is (= (page-text browser) "Sparkledriver, driven\n\nAnd what rough beast, its hour come round at last\n\nSpiritus Mundi\n\nThe blood-dimmed tide is loosed, and everywhereThe ceremony of innocence is drowned"))
      (is (= (title browser) "The Second Coming")))

    (testing "cookies"
      (is (= (-> browser
                 (set-cookie! "no-domain" "uses current domain")
                 (set-cookie! "no-domain" "can be changed")
                 (set-cookie! "with-domain" "SparkleDriver" :domain "example.org" :path "/home")
                 browser-cookies->map)
             {:no-domain   {:domain "0.0.0.0", :path "/", :value "can be changed"},
              :with-domain {:domain "example.org", :path "/home", :value "SparkleDriver"}}))

      (is (= (-> browser
                 delete-all-cookies!
                 browser-cookies->map) {}))

      (is (= (-> browser
                 (set-cookie! "a-cookie" "no domain" :path "/foo")
                 (set-cookie! "a-cookie" "no domain" :path "/bar")
                 (set-cookie! "a-cookie" "some domain" :domain "example.org")
                 (set-cookie! "a-cookie" "some domain" :domain "example.org" :http-only true, :secure true, :path "/foo")
                 (set-cookie! "a-cookie" "some other domain" :domain "clojure.org")
                 (.. manage getCookies)
                 (->> (map str))
                 sort)
             ["a-cookie=no domain; path=/bar; domain=0.0.0.0"
              "a-cookie=some domain; path=/foo; domain=example.org;secure;"
              "a-cookie=some other domain; path=/; domain=clojure.org"])))))

(deftest fetch-test
  (with-browser [browser (make-browser)]
    (testing "it coerces the URL to string"
      (let [url (str "http://0.0.0.0:" port "/the/path")]
        (is (= (current-url (fetch! browser (URI. url))) url))))))

(deftest retry-test
  (testing "retry success"
    (is (= "success")
        (let [n (atom 1)]
          (with-retry (partial retry-backoff 8 (fn [_] (swap! n dec)))
            (*retry-fn* 
             #(if (= 0 @n)
                "success"
                (throw (Exception. "fail!"))))))))
  (testing "retry failure"
    (is (= "fail!"
           (try
             (with-retry (partial retry-backoff 4)
               (*retry-fn* #(throw (Exception. "fail!"))))
             (catch Exception e "fail!"))))))

(deftest close-test
  (with-browser [browser (fetch! (make-browser) (str "http://0.0.0.0:" port "/"))]
    (testing "can close the browser with multiple windows open"
      (execute-script browser (str "window.open('http://0.0.0.0:" port "');"))
      (switch-to-window browser "1")
      (find-by-tag browser "h1"))))

(deftest block-ads
  (let [js "function isAdLoaded(img) { if (typeof img === 'undefined') { return true; }; if (typeof img.naturalWidth !== 'undefined' && img.naturalWidth <= 1) { return true; }; return false; }\n
            document.getElementById('test-ad').src = 'http://widgets.outbrain.com/images/widgetIcons/ob_logo_16x16.png?advertiser=1&' + escape(new Date());\n
            return isAdLoaded(document.getElementById('test-ad'));"]
    (testing "ad should get blocked"
      (with-browser [browser (fetch! (make-browser :block-ads true) (str "http://0.0.0.0:" port))]
        (is (= (execute-script browser js)
               true))))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; for quick interactive tests

;;(def browser (make-browser))

;; (page-source browser)

;; (start-server port)
;; (with-browser [browser (fetch! (make-browser :screen-size [1900 1000]) (str "http://0.0.0.0:" port))]
;;   (-> (find-by-tag browser "h1") inner-html)
;;   )


project.clj
(defproject sparkledriver "0.2.4"
  :description "A clojure wrapper for jBrowserDriver, which is a Selenium-compatible wrapper around JFX embedded WebKit."
  :url "https://github.com/jackrusher/sparkledriver"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.9.0"]
                 [com.machinepublishers/jbrowserdriver "1.0.1"]
                 [org.clojure/math.numeric-tower "0.0.4"]]
  :profiles {:dev {:dependencies [[org.slf4j/slf4j-simple "1.7.25"]
                                  [http-kit "2.3.0"]
                                  [compojure "1.6.1"]
                                  [hiccup "1.0.5"]]}
             :codox {:dependencies [[codox-theme-rdash "0.1.2"]]
                     :plugins [[lein-codox "0.10.3"]]
                     :codox {:project {:name "sparkledriver"}
                             :metadata {:doc/format :markdown}
                             :themes [:rdash]
                             :output-path "gh-pages"}}}
  :aliases {"codox" ["with-profile" "codox,dev" "codox"]}
  :deploy-repositories [["releases" :clojars]])


src/sparkledriver/retry.clj
(ns sparkledriver.retry
  (:require [clojure.math.numeric-tower :refer [expt]]))

(defonce ^:dynamic *retry-fn* (fn [f] (f)))

(defn retry-backoff
  "Execute `try-fn`, catching any exception and then retrying with an exponential backoff until a retry pause of `max-wait` seconds is reached. If `recover-fn` is supplied, it will be called before each retry with the most recent exception as its only argument. Either returns the result of a successful invocation of `try-fn` or raises an exception.

  If try-fn throws an `:unhandled-fatal` exception the retry loop will terminate.

  Simple example:
  
```
  (retry-backoff
    ;; retry at 2,4 and 8 seconds
    8
    ;; return the value of element w/ id \"the-id\" if successful
    #(find-by-id browser \"the-id\"))
```

  Add a `recovery-fn` to the retry, which in this case just logs the exception:

```
  (retry-backoff
    ;; retry at 2,4,8 and 16 seconds
    16
    ;; \"recover-fn\" that logs failures between retries
    #(info (str \"Trying again after a \" (.getMessage %) \" exception...\"))
    ;; return the value of element w/ id \"the-id\" if successful
    #(find-by-id browser \"the-id\"))
```
  "
  ([max-wait try-fn] (retry-backoff max-wait nil try-fn))
  ([max-wait recover-fn try-fn]
   (loop [retry 0]
     (let [[success result ex] (try
                                 [true (try-fn) nil]
                                 (catch Exception e [false nil e]))]
       (if success
         result
         (let [x-info (ex-data ex)
               wait (+ (expt 2 retry) 2)]        ; 2s min wait
           (when (= (:cause x-info) :unhandled-fatal) ; bubble up :unhandled-fatal to escape retries
             (throw ex))
           (if (> wait max-wait)
             (throw (Exception. (str (.getMessage ex) " - timeout exceeded, too many retries!")))
             (do
               (when recover-fn
                 (recover-fn ex))
               (Thread/sleep (* 1000 wait))
               (recur (inc retry))))))))))

(defmacro with-retry
  "Evaluate `body` with `retry-fn` bound to the dynamic variable `*retry-fn*`. Many Sparkledriver functions automatically wrap themselves in `*retry-fn*` when it is bound, which allows one to specify a retry policy for those functions within the scope of this macro.

  Example:

```
  (with-browser [browser (make-browser)]
    (fetch! browser \"http://clojure.org\")  
    (with-retry (partial retry-backoff 16)
      (-> browser
          (find-by-xpath* \"//div[@class='clj-intro-message']/p\")
          (nth 2)
          text)))
;;=> \"I hope you find Clojure's combination of facilities elegant, powerful, practical and fun to use.\"      
```"
  [retry-fn & body]
  `(binding [*retry-fn* ~retry-fn]
     ~@body))



src/sparkledriver/cookies.clj
(ns sparkledriver.cookies
  (:require [sparkledriver.browser :as brwsr]))

(defn browser-cookies->map
  "Convert `browser`'s current cookies into the map format used by clj-http. This returns cookies from all domains, but cookie names that are set on multiple domains will only be returned once."
  [browser]
  (reduce
   #(assoc %1 (keyword (.getName %2))
           {:domain (.getDomain %2)
            :path   (.getPath %2)
            :value  (.getValue %2)})
   {}
   (.getCookies (.manage browser))))

(defn delete-all-cookies!
  "Clear all cookies from all domains."
  [browser]
  (.deleteAllCookies (.manage browser))
  browser)

(defn delete-cookie!
  "Delete the named cookie from the given domain. When omitted `domain` defaults to the `browser`'s current domain."
  ([browser name]
   (delete-cookie! browser name (.getHost (java.net.URL. (brwsr/current-url browser)))))
  ([browser name domain]
   (let [manager (.manage browser)
         cookies (.getCookies manager)]
     (when-let [cookie (some (fn [c]
                               (and (= (.getName c) name)
                                    (= (.getDomain c) domain)
                                    c))
                             cookies)]
       (.deleteCookie manager cookie))
     browser)))

(defn- build-selenium-cookie
  "Build an instance of org.openqa.selenium.Cookie with `name`, `value` and `options`."
  [name value options]
  (.build
   (reduce
    (fn [builder [k v]]
      (case k
        :domain (.domain builder v)
        :expires-on (.expiresOn builder v)
        :http-only (.isHttpOnly builder v)
        :secure (.isSecure builder v)
        :path (.path builder v)
        builder))
    (org.openqa.selenium.Cookie$Builder. name value)
    options)))

(defn set-cookie!
  "Set a cookie with given `name` and `value`. If a cookie with the same name and `domain` is already present, it will be replaced. The following keyword arguments are supported.

  - `:domain` (String) The cookie domain, defaults to the current domain of the browser.
  - `:expires-on` (java.util.Date) Expiration date. Must be in the future or the cookie won't be stored.
  - `:http-only` (Boolean) Cookie's HttpOnly flag, disallow access from JavaScript
  - `:secure` (Boolean) Cookie's Secure flag, only serve over HTTPS.
  - `:path` (String) URL path of the cookie."
  [browser name value & {domain :domain :as options}]
  (if domain
    (delete-cookie! browser name domain)
    (delete-cookie! browser name))
  (.addCookie (.manage browser) (build-selenium-cookie name value options))
  browser)


src/sparkledriver/browser.clj
(ns sparkledriver.browser
  (:require [sparkledriver.element :as elem]))

(def browser-options
  "The possible options for building a browser instance in {:option [default setter-fn]} format."
  {;; how long to wait for resources loaded by ajax, in milliseconds (default is quite long)
   :ajax-load-timeout [30000 #(.ajaxResourceTimeout %1 %2)]
   ;; how long to wait for JS to run after page load, in milliseconds
   :ajax-wait         [200 #(.ajaxWait %1 %2)]
   ;; whether requests to ad/spam servers should be blocked
   :block-ads         [false #(.blockAds %1 %2)]
   ;; use a local browser cache
   :cache             [true #(.cache %1 %2)]
   ;; if false, open a window so you can watch it work
   :headless          [true #(.headless %1 %2)]
   ;; increased logging
   :log-trace         [false #(.logTrace %1 %2)]
   :log-wire          [false #(.logWire %1 %2)]
   ;; we pretend to be Chrome
   :request-headers   [com.machinepublishers.jbrowserdriver.RequestHeaders/CHROME #(.requestHeaders %1 %2)]
   ;; store copies of media and attachments in a temporary folder
   :save-attachments  [true #(.saveAttachments %1 %2)]
   :save-media?       [true #(.saveMedia %1 %2)]
   ;; set browser screen dimentions - 1366x768 by default (we're a laptop)
   :screen-size       [[1366 768] (fn [builder [w h]]
                                    (.screen builder (org.openqa.selenium.Dimension. w h)))]
   ;; be accepting of weird SSL certs
   :ssl-policy        ["compatible" #(.ssl %1 %2)]
   ;; We're in New York, no matter where we are
   :timezone          [com.machinepublishers.jbrowserdriver.Timezone/AMERICA_NEWYORK #(.timezone %1 %2)]
   ;; no, really, we're Chrome
   :user-agent        [com.machinepublishers.jbrowserdriver.UserAgent/CHROME #(.userAgent %1 %2)]
   ;; SSL certificate verification, off by default because the internet is broken
   :verify-hostname?  [false #(.hostnameVerification %1 %2)]})

(defn make-browser
  "Creates a new headless browser instance parameterized by `options`. Examples:

```clojure
  (make-browser :log-wire true)  ; log network traffic

  (make-browser :ajax-wait 1000) ; wait 1000ms for JS to run after page load

  (make-browser :headless false) ; pop up a browser window to watch it work
```
  "
  [& options]
  (assert (or (= nil options) (even? (count options)))
          "The options to make-browser must be an even number of key-value pairs.")
  (let [merged-opts (->> (partition 2 options)
                         (reduce (fn [a [k v]]
                                   (if-let [[default setter-fn] (get browser-options k)]
                                     (assoc a k [v setter-fn])
                                     (throw (IllegalArgumentException. "Invalid browser option."))))
                                 browser-options)
                         vals)]
    (com.machinepublishers.jbrowserdriver.JBrowserDriver.
     (.build (reduce (fn [builder [v f]]
                       (f builder v))
                     (com.machinepublishers.jbrowserdriver.Settings/builder)
                     merged-opts)))))

(defn close-browser!
  "Close a browser instance, killing the underlying subprocess and freeing all resources."
  [browser]
  ;; upstream bug causes crash when not on the first window when there
  ;; are multiple open windows
  (let [handles (.getWindowHandles browser)]
    (when (> (count handles) 1)
      (.window (.switchTo browser) (first handles))))
  (.quit browser))

(defmacro with-browser
  "Evaluate `body` in a try block within which `binding` is bound, finally calling close-browser on binding. This is just a version of with-open that traps the exception that closing a jBrowserDriver instance currently throws."
  [binding & body]
  ;; assert-args is private to clojure.core 😿
  ;; (assert-args
  ;;    (vector? binding) "a vector for its binding"
  ;;    (= 1 (count binding)) "a single name is expected"
  ;;    (symbol? (binding 0)) "the name should be a symbol")
  `(let ~(subvec binding 0 2)
     (try
       ~@body
       (finally (close-browser! ~(binding 0))))))

(defn fetch!
  "Fetch 'url' using 'browser' and return browser after loading is complete.

```clojure
  (fetch! browser \"http://www.clojure.com/\")
  ;;=> returns browser after navigating to the clojure site
```"
  [browser url]
  (.get browser (str url))
  browser)

(defn status-code
  "Return the current HTTP status code of `browser`."
  [browser]
  (.getStatusCode browser))

(defn current-url
  "Return the currently visited url of `browser`."
  [browser]
  (.getCurrentUrl browser))

(defn page-source
  "Return the HTML source of the currently visited page in `browser`."
  [browser]
  (.getPageSource browser))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; browser windows

(defn current-window
  "Return a handle for the currently active `browser` window."
  [browser]
  (.getWindowHandle browser))

(defn all-windows
  "Return a set of handles for all `browser` windows."
  [browser]
  (.getWindowHandles browser))

(defn switch-to-window
  "Switch `browser` to `window`, which should be a handle returned by all-windows or current-window."
  [browser window]
  (.window (.switchTo browser) window))

(defn maximize-window
  "Maximize `browser`'s currently focused window."
  [browser]
  (.maximize (.window (.manage browser))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; alert handling

(defn switch-to-alert
  "Switch `browser`'s focus to the current alert popup. Returns the alert's handle."
  [browser]
  (.alert (.switchTo browser)))

(defn accept-alert!
  "Accept an alert."
  [alert]
  (.accept alert))

(defn dismiss-alert!
  "Dismisses an alert."
  [alert]
  (.dismiss alert))

;; XXX currently locks up and never comes back! Upstream problem in
;; jBrowserDriver?
;;
;; (defn alert-text
;;   "Returns the text of an alert."
;;   [alert]
;;   (.getText alert))

;; TODO add typing into alerts...
;;void	sendKeys(String keysToSend) 

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; exec arbitrary javascript

(defn execute-script
  "Executes JavaScript `script` in the context of the currently selected frame or window of `browser` with `arguments`."
  [browser script & arguments]
  (.executeScript browser script (to-array arguments)))

(defn execute-script-async
  "Asynchronously execute JavaScript `script` in the context of the currently selected frame or window of `browser` with `arguments`."
  [browser script & arguments]
  (.executeAsyncScript browser script (to-array arguments)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; interrogate logs

(defn available-log-types
  "Return a set of strings representing the types of logs available for `browser`. N.B. `browser` must have been initialized with some kind of logging enabled, for isntance :log-trace or :log-wire."
  [browser]
  (.getAvailableLogTypes (.logs (.manage browser))))

(defn logs
  "Return `browser`'s logs of whatever `kind`, or all if not specified. See http://machinepublishers.github.io/jBrowserDriver/org/openqa/selenium/logging/LogEntry.html for details of the returned items. Example:

```
  (map #(.getMessage %) (logs browser))
  ;;=> events that have occurred since the last time you did this
```"
  [browser & kind]
  (iterator-seq (.iterator (.get (.logs (.manage browser)) (or kind "all")))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; helpers

(defn page-wait
  "Blocks until the browser's current page quiesces. N.B. This is the default for many operations, so this function is likely only useful when things are misbehaving."
  [browser]
  (.pageWait browser))

(defn attachments-dir
  "Return the path to the directory where the browser is storing attachment files."
  [browser]
  (.attachmentsDir browser))

(defn media-dir
  "Return the path to the directory where the browser is storing media files."
  [browser]
  (.mediaDir browser))

(defn cache-dir
  "Return the path to the directory where the browser is storing cached files."
  [browser]
  (.cacheDir browser))

(defn page-text
  "Return the complete visible textual content of the current page in the focused window of `browser`. Text from hidden elements is not included."
  [browser]
  (elem/text (elem/find-by-tag browser "html")))

(defn title
  "Return the title of the current page in the focused window of `browser`."
  [browser]
  (elem/inner-html (elem/find-by-css browser "head title")))
